proposals/nnnn-variadic-types.md

```diff
@@ -0,0 +1,239 @@
+
+  let seq: (repeat each S)
+
+  func makeIterator() -> Iterator {
+    return Iterator(iter: (repeat (each seq).makeIterator()))
+  }
+
+  struct Iterator: IteratorProtocol {
+    typealias Element = (repeat each S.Element)
+
+    var iter: (repeat each S.Iterator)
+
+    mutating func next() -> Element? {
+      return ...
+    }
+  }
+}
+```
+
+## Proposed solution
+
+In the generic parameter list of a generic type, the `each` keyword declares a generic parameter pack, just like it does in the generic parameter list of a generic function. The types of stored properties can contain pack expansion types, as in `let seq` and `var iter` above.
+
+## Detailed design
+
+Swift has the following kinds of generic type declarations:
+
+- Structs
+- Enums
+- Classes (and actors)
+- Type aliases
+
+A generic type is _variadic_ if it directly declares a type parameter pack with `each`, or if it is nested inside of another variadic type. In this proposal, structs, classes, actors and type aliases can be variadic. Enums will be addressed in a follow-up proposal.
+
+### Single parameter
+
+A generic type is limited to declaring at most one type parameter pack. The following are allowed:
+
+```swift
+struct S1<each T> {}
+struct S2<T, each U> {}
+struct S3<each T, U> {}
+```
+
+But this is not:
+
+```swift
+struct S4<each T, each U> {}
+```
+
+However, by virtue of nesting, a variadic type might can still abstract over multiple type parameter packs:
+
+```swift
+struct Outer<each T> {
+  struct Inner<each U> {
+    var fn: (repeat each T) -> (repeat each U)
+  }
+}
+```
+
+### Referencing a variadic type
+
+When used with a variadic type, the generic argument syntax `S<...>` allows a variable number of arguments to be specified. Since there can only be one generic parameter pack, the non-pack parameters are always specified with a fixed prefix and suffix of the generic argument list.
+
+```swift
+struct S<T, each U, V> {}
+
+S<Int, Float>.self  // T := Int, U := Pack{}, V := Float
+S<Int, Bool, Float>.self  // T := Int, U := Pack{Bool}, V := Float
+S<Int, Bool, String, Float>.self  // T := Int, U := Pack{Bool, String}, V := Float
+```
+
+Note that `S<Int, Float>` substitutes U with the empty pack type, which is allowed. The minimum number of generic arguments is equal to the number of non-pack generic parameters. In our above example, the minimum argument count is 2, because `T` and `V` must always be specified:
+
+```swift
+S<Int>.self // error: expected at least 2 generic arguments
+```
+
+If the generic parameter list of a variadic type consists of a single generic parameter pack and nothing else, it is possible to reference it with an empty generic argument list:
+
+```swift
+struct V<each T> {}
+
+V<>.self
+```
+Note that `V<>` is not the same as `V`. The former substitutes the generic parameter pack `T` with the empty pack. The latter does not constrain the pack at all and is only permitted in contexts where the generic argument can be inferred.
```

Comment:
This is sort of covered under 'inferred' I guess, but maybe worth being explicit about the 'bare name means `Self` within the generic type' rule here (I don't think this language is intended to _ban_ that behavior inside variadic generic types, right?)